core/src/main/java/org/apache/calcite/sql/type/SetopOperandTypeChecker.java

```diff
@@ -92,27 +93,44 @@ public boolean checkOperandTypes(
     // column j.
     for (int i = 0; i < colCount; i++) {
       final int i2 = i;
+
+      List<RelDataType> columnIthTypes = new AbstractList<RelDataType>() {
+        public RelDataType get(int index) {
+          return argTypes[index].getFieldList().get(i2)
+              .getType();
+        }
+
+        public int size() {
+          return argTypes.length;
+        }
+      };
+
       final RelDataType type =
-          callBinding.getTypeFactory().leastRestrictive(
-              new AbstractList<RelDataType>() {
-                public RelDataType get(int index) {
-                  return argTypes[index].getFieldList().get(i2)
-                      .getType();
-                }
-
-                public int size() {
-                  return argTypes.length;
-                }
-              });
+          callBinding.getTypeFactory().leastRestrictive(columnIthTypes);
       if (type == null) {
-        if (throwOnFailure) {
-          SqlNode field =
-              SqlUtil.getSelectListItem(callBinding.operand(0), i);
-          throw validator.newValidationError(field,
-              RESOURCE.columnTypeMismatchInSetop(i + 1, // 1-based
-                  callBinding.getOperator().getName()));
-        } else {
-          return false;
+        boolean changed = false;
+        if (validator.getEnableTypeCoercion()) {
+          for (int j = 0; j < callBinding.getOperandCount(); j++) {
+            TypeCoercion typeCoercion = validator.getTypeCoercion();
+            RelDataType widenType = typeCoercion.getWiderTypeFor(columnIthTypes, true);
+            if (null != widenType) {
+              changed = typeCoercion.widenColumnTypes(
+                  callBinding.getScope(),
+                  callBinding.operand(j), i, widenType)
+                  || changed;
```

Comment:
why need make or ?

Reply:
I do it for the first decision just to sync the code pattern, we can remove it.